### ISSUE TITLE
Database Class 

### ISSUE THREAD

**Issue by Arjun-NA** [HuguesTHOMAS/KPConv-PyTorch#37]
I currently have my own code for generating batches of points and labels as below:

 ```

def load_lasfile(filename):
    inFile = laspy.file.File(filename)
    data = np.vstack([inFile.x, inFile.y, inFile.z, inFile.classification]).transpose()
    points = data.astype(np.float32)
    # labels_seg = data.loc[:, 'class'].to_numpy().astype(np.int64)
    return points  # labels_seg


def minMax(x):
    return pd.Series(index=['min', 'max'], data=[x.min(), x.max()])


class Dataset(IterableDataset):
    """
    """

    def __init__(self, file_in, path='../Data/DALES', batch_size=32, npoints=2048):
        super(Dataset).__init__()
        self.file_list = [os.path.join(path, line.strip()) for line in open(os.path.join(path, file_in))]
        self.batch_size = batch_size
        self.npoints = npoints
        self.index = 0
        self.first_batch = True
        self.sq_size = 0
        self.label_to_names = {0: 'unknown',
                               1: 'buildings',
                               2: 'cars',
                               3: 'trucks',
                               4: 'poles',
                               5: 'power lines',
                               6: 'fences',
                               7: 'ground',
                               8: 'chair',
                               9: 'vegetation'}
        self.init_labels()
        self.ignored_labels = np.array([0])
        self.dataset_task = 'segmentation'

    def next_batch(self):
        # __file = self.file_list[self.index]
        for each_file in self.file_list:
            # Find sq_size which gives value of x and y coordinates skip considering uniformity in data files
            batch_accumulation = 0
            data_ = load_lasfile(each_file)
            print("-----------{}-----------".format(each_file))
            if self.first_batch:
                data_min = data_.min(axis=0)
                data_max = data_.max(axis=0)
                x_range = data_max[0] - data_min[0]
                y_range = data_max[1] - data_min[1]
                x_unq = np.sort(np.unique(data_[:, 0]))
                y_unq = np.sort(np.unique(data_[:, 1]))
                size = 0
                while size < self.npoints and self.first_batch:
                    self.sq_size += 1
                    pos_x = np.logical_and(x_unq[self.sq_size] > data_[:, 0],
                                           data_[:, 0] >= x_unq[0])
                    pos_y = np.logical_and(y_unq[self.sq_size] > data_[:, 1],
                                           data_[:, 1] >= y_unq[0])
                    pos = np.logical_and(pos_x, pos_y)
                    size = len(data_[pos])
                self.sq_size -= 1  # to include not more than self.npoints
            self.first_batch = False
            self.y_ll = 0
            while (self.y_ll + self.sq_size) < len(y_unq):
                self.y_ul = self.y_ll + self.sq_size
                self.x_ll = 0
                while (self.x_ll + self.sq_size) < len(x_unq):
                    self.x_ul = self.x_ll + self.sq_size
                    pos_x = np.logical_and(x_unq[self.x_ul] > data_[:, 0],
                                           data_[:, 0] >= x_unq[self.x_ll])
                    pos_y = np.logical_and(y_unq[self.y_ul] > data_[:, 1],
                                           data_[:, 1] >= y_unq[self.y_ll])
                    pos = np.logical_and(pos_x, pos_y)
                    self.x_ll += self.sq_size
                    data = data_[pos]
                    data = np.resize(data, (self.npoints, data_.shape[1]))
                    if labels_seperate:
                        if batch_accumulation < self.batch_size:
                            if batch_accumulation == 0:
                                data_batch = np.reshape(data[:, :-1], (1, data.shape[0], data.shape[1] - 1))
                                labels_batch = np.reshape(data[:, -1], (1, -1))
                            else:
                                data_batch = np.vstack(
                                    (data_batch, data[:, :-1].reshape((1, data.shape[0], data.shape[1] - 1)))
                                )
                                labels_batch = np.vstack(
                                    (labels_batch, data[:, -1].reshape((1, -1)))
                                )
                            batch_accumulation += 1
                        else:
                            batch_accumulation = 0
                            yield data_batch, labels_batch
                    else:
                        if batch_accumulation < self.batch_size:
                            if batch_accumulation == 0:
                                data_batch = np.reshape(data, (1, data.shape[0], data.shape[1]))
                            else:
                                data_batch = np.vstack(
                                    (data_batch, data.reshape((1, data.shape[0], data.shape[1])))
                                )
                            batch_accumulation += 1
                        else:
                            batch_accumulation = 0
                            yield data_batch
                self.y_ll += self.sq_size
            self.y_ul = -1
            self.x_ul = -1
            pos_x = np.logical_and(x_unq[self.x_ul] > data_[:, 0],
                                   data_[:, 0] >= x_unq[self.x_ll])
            pos_y = np.logical_and(y_unq[self.y_ul] > data_[:, 1],
                                   data_[:, 1] >= y_unq[self.y_ll])
            pos = np.logical_and(pos_x, pos_y)
            data = data_[pos]
            data = np.resize(data, (self.npoints, data_.shape[1]))
            if labels_seperate:
                if batch_accumulation < self.batch_size:
                    if batch_accumulation == 0:
                        data_batch = np.reshape(data[:, :-1], (1, data.shape[0], data.shape[1] - 1))
                        labels_batch = np.reshape(data[:, -1], (1, -1))
                    else:
                        data_batch = np.vstack(
                            (data_batch, data[:, :-1].reshape((1, data.shape[0], data.shape[1] - 1)))
                        )
                        labels_batch = np.vstack(
                            (labels_batch, data[:, -1].reshape((1, -1)))
                        )
                    yield data_batch, labels_batch
            else:
                if batch_accumulation < self.batch_size:
                    if batch_accumulation == 0:
                        data_batch = np.reshape(data, (1, data.shape[0], data.shape[1]))
                    else:
                        data_batch = np.vstack(
                            (data_batch, data.reshape((1, data.shape[0], data.shape[1])))
                        )
                yield data_batch

    def init_labels(self):
        # Initialize all label parameters given the label_to_names dict
        self.num_classes = len(self.label_to_names)
        self.label_values = np.sort([k for k, v in self.label_to_names.items()])
        self.label_names = [self.label_to_names[k] for k in self.label_values]
        self.label_to_idx = {l: i for i, l in enumerate(self.label_values)}
        self.name_to_label = {v: k for k, v in self.label_to_names.items()}

    def __iter__(self):
        return self.next_batch()

    def __len__(self):
        return 2000
```

class Dataset was imported as DalesDataset and used in train file as below.

```
    training_dataset = DalesDataset(config.train_filelist, path='../Data/DALES/train', batch_size=1, npoints=2048)
    test_dataset = DalesDataset(config.test_filelist, path='../Data/DALES/test', batch_size=1, npoints=2048)

    training_loader = DataLoader(training_dataset,
                                 batch_size=1,
                                 num_workers=0, 
                                 pin_memory=True)
    test_loader = DataLoader(test_dataset,
                             batch_size=None,
                             num_workers=0, 
                             pin_memory=True)
```
but I run into many errors. How should the data look to feed into your model?

I did not understand the length for the generator. so I added a random 2000 for checking how it runs.
But I find it hard to integrate this into your code.  Is it possible to use this code or should I make changes?
I just need an idea or a link from where I can make it proper.

Thanks,
Arjun

**Comment by HuguesTHOMAS** [comment]
I just learned PyTorch while coding this repo, so I just followed tutorials on PyTorch website ([here ](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html)or [here](https://pytorch.org/docs/stable/data.html)) and everytime I had a problem, I found a solution on stackoverflow or similar forums.

About the length, it is useless, the code never calls it, so you don't have to worry about it.

I would advise not to use iterable dataset as they are not very versatile. Prefer map style datasets and try to take inspiration from the dataset I already coded like S3DIS.

**Comment by Arjun-NA** [comment]
Hi @HuguesTHOMAS  Thank you for your reply. I have one problem using the S3DIS as example, in it you have mentioned the data is small enough to be loaded into the memory. For me it is not so is it still feasible to use the same structure? Thanks.

**Comment by HuguesTHOMAS** [comment]
If you have a lot of small files (like lidar scans or RGBD images), you can use SemanticKitti dataset as an example. Otherwise if you have large scenes it is going to be much more difficult for you because I did not code anything for that case.

I kinda thought of this problem and here is my advise on it. If you have a lot of very large scenes, you need to break down your dataset in small tiles (each tile in a separate file) so that you can load the points fast. Then when you need a new input cloud (lets imagine a sphere like I do in S3DIS), you need to compute which tiles your sphere intersect, load them, and then get the points in the sphere. This will not be simple to implement but it should be good to keep a fast training.

**Comment by Arjun-NA** [comment]
Okay, Mine is Lidar point cloud which have in total of 0.5 B points but it is tiled into smaller squares. I think I need to understand your S3DIS code more. Anyways, do you find using spherical neighbors highly impacting accuracy? I thought of making small square area with input points just more than ( let's say ) 4096 points and then reduce the size to 4096 to make it equal for all iterations. I read that (from the paper) you have implemented something pretty good to consider the variation in the input points. I am yet to find that snip of your code. Or is it such that I don't have to create a stable number of input points. The network would adapt to it?

Thank you so much for taking your time to reply my questions

**Comment by HuguesTHOMAS** [comment]
> Or is it such that I don't have to create a stable number of input points. The network would adapt to it?

Yes it is like that. 

First the network does not need a constant number of point as input like Pointnet. The input number of points is variable. It is possible because we use "stacked" batch instead of regular ones. We don't have a batch with shape `[B, N, ...]` but we stack all clouds like this: `[N1 + N2 + N3, ...]`, therefore N1, N2, and N3 do not have to be the same.

Second, to have a relatively stable memory consumption, the network automatically adapts the batch size depending on the size of the input clouds. Let's say you ask for `batch_num = 8`. The network is not going to always use 8 input clouds in each batch. Instead it will add clouds to the batch until it reaches a certain limit `N_max`. Sometimes the batch will contain 5 big clouds, and some other times it will contain 10 smaller clouds but in average it will be around 8. You can see this in the calibration function of S3DIS dataset:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/94b1d87e17ec339e2aad049ce5d5d175b9cd0db1/datasets/S3DIS.py#L1038

**Comment by xdw-123** [comment]
@Arjun-NA Hello, has your problem been solved? I found the same dataset, but didn't know how to train the model?

**Comment by Arjun-NA** [comment]
@xdw-123  Hi , I used tensorflow version for now, in which NPM3d dataset code is easily convertible to this dataset. If I have any progress with pytorch I will update it soon here.

**Comment by xdw-123** [comment]
> @xdw-123 Hi , I used [tensorflow version](https://github.com/HuguesTHOMAS/KPConv) for now, in which [NPM3d](https://github.com/HuguesTHOMAS/KPConv/blob/master/datasets/NPM3D.py) dataset code is easily convertible to this dataset. If I have any progress with pytorch I will update it soon here.

Thank you for your reply. Do you mean that dataset DALES can be converted to this required dataset format through NPM3d dataset code?

**Comment by Arjun-NA** [comment]
@xdw-123  You can download the DALES in ply format which is available with the mail they send you
![image](https://user-images.githubusercontent.com/21004258/94099661-7df34380-fdf9-11ea-8e56-492900f77a0f.png)
This location can be then set in the database file. no need of any other conversions.

**Comment by xdw-123** [comment]
> @xdw-123 You can download the DALES in ply format which is available with the mail they send you
> ![image](https://user-images.githubusercontent.com/21004258/94099661-7df34380-fdf9-11ea-8e56-492900f77a0f.png)
> This location can be then set in the database file. no need of any other conversions.

Thank you for your reply. I'll try it right away.There are three formats for the same dataset in the pictures you send. Do you want only one?Is it necessary to have all three?

**Comment by xdw-123** [comment]
@Arjun-NA Hello, I tried running Python 3.5 training_NPM3D.py.Many problems have arisen.But most of them are read and write errors in ply files. Can you upload your modifications?

**Comment by Arjun-NA** [comment]
@xdw-123 sorry for the late reply.
I missed out that the ply file that they provide is in ascii format and you would need in binary format. I will upload the conversion file in here.

**Comment by eyalol** [comment]
Hello @Arjun-NA, @xdw-123 , I've managed to train RandLA-Net on DALES with [this](https://github.com/aRI0U/RandLA-Net-pytorch/blob/master/data.py), I'm currently investigating what do I need to change in this dataset to fit KPConv. I'm assuming you could use it too for the DALES dataset.

Best Regards,

**Comment by Arjun-NA** [comment]
Hi @eyalol , I am not sure about the dataset RandLA-Net, please check with Dr. Hughes for more information. Thanks
